zerver/actions/message_edit.py

```diff
@@ -1576,3 +1578,30 @@ def check_update_message(
             notify_stream_is_recently_active_update(new_stream, is_stream_active)
 
     return updated_message_result
+
+
+@transaction.atomic(durable=True)
+def re_thumbnail(
+    message_class: type[Message] | type[ArchivedMessage], message_id: int, enqueue: bool
+) -> None:
+    message = message_class.objects.select_for_update().get(id=message_id)
+    assert message.rendered_content is not None
+    image_metadata = get_user_upload_previews(
+        message.realm_id,
+        message.rendered_content,
+        enqueue=enqueue,
+        lock=True,
+    )
+
+    new_content, _ = rewrite_thumbnailed_images(
+        message.rendered_content,
+        image_metadata,
+    )
+
+    if new_content is None:
+        pass
+    elif isinstance(message, Message):
+        do_update_embedded_data(message.sender, message, new_content)
+    else:
+        message.rendered_content = new_content
+        message.save(update_fields=["rendered_content"])
```

Comment:
This might be more readable to have a `assert isinstance(message, ArchivedMessage);` and/or a comment noting that we just need to update deleted messages in the database.